fdbserver/ptxn/TLogServer.actor.cpp

```diff
@@ -443,21 +492,71 @@ struct LogGenerationData : NonCopyable, public ReferenceCounted<LogGenerationDat
 		StorageTeamID storageTeamId;
 		std::vector<Tag> tags;
 		std::map<Version, std::pair<StringRef, Arena>> versionMessages;
-
-		StorageTeamData(StorageTeamID storageTeam, std::vector<Tag> tags) : storageTeamId(storageTeam), tags(tags) {}
+		bool nothingPersistent; // true means tag is *known* to have no messages in persistentData.  false means nothing.
+		bool poppedRecently; // `popped` has changed since last updatePersistentData
+		Version popped; // see popped version tracking contract below
+		Version persistentPopped; // The popped version recorded in the btree.
+		Version versionForPoppedLocation; // `poppedLocation` was calculated at this popped version
+		IDiskQueue::location poppedLocation; // The location of the earliest commit with data for this tag.
+		bool unpoppedRecovered;
+
+		StorageTeamData(StorageTeamID storageTeam, std::vector<Tag> tags)
+		  : storageTeamId(storageTeam), tags(tags), popped(0), poppedLocation(0), persistentPopped(0),
+		    versionForPoppedLocation(0), poppedRecently(false), unpoppedRecovered(false), nothingPersistent(nothingPersistent) {}
+
+		StorageTeamData(StorageTeamID storageTeam, std::vector<Tag> tags, Version popped)
+		  : storageTeamId(storageTeam), tags(tags), popped(popped), poppedLocation(0), persistentPopped(0),
+		    versionForPoppedLocation(0), poppedRecently(false), unpoppedRecovered(false), 
+			nothingPersistent(false) {}
```

Comment:
Nit: Prefer to initialize members where they are declared.

Reply:
will fix